### PR TITLE
fabtests/runfabtests.py: Pytest based runfabtests

### PR DESCRIPTION
**Pull request by wzamazon**
This patch introduced a set of pytest based scripts that re-implements
runfabtests.sh in python.
    
It consists of the following scripts:
    
    runfabtests.py: a python script that accepts the same set of arguments
                    runfabtests.sh does.
    
    pytest/common.py: provides utility classes like UnitTest, ClientServerTest,
                      MultinodeTest and ComplexTest
    
    pytest/conftest.py: set up pytest fixtures like client_id, server_id and
                        command_line_args
    
    pytest/default: a set of pytest test cases that cover all tests defined
                    in runfabtests.sh

It also introduce some new functionality that runfabtests.sh does not support:

1. it provides a way for each provide to define its own set of tests.
2. it proviers a way to handle combinations of parameters for each test by using pytest's paramter marks.
3. it provides an option for user to choose a set of tests by regex expression
4. it provides an option to generate a html format test report to be used by jenkins server
5. it provides an option to generate a junit xml format test report to be used by jenkins server.

    
    Signed-off-by: Wei Zhang <wzam@amazon.com>

### CODE REVIEW

**Comment by shefty** [comment]
Nice!  Instead of using the term complex, can we just call out the one test this refers to, ubertest?  I'd like to consider aligning on the test names based on how they're organized: benchmarks, unit, functional, multinode, and ubertest.  It looks like this is mostly there.

Is the plan to eventually remove the runfabtests.sh script?  (I'd prefer a single set of scripts to maintain).

**Comment by wzamazon** [comment]
>Nice! Instead of using the term complex, can we just call out the one test this refers to, ubertest? 

Sure, I used the term "complex" because it is the term used by "runfabtests.sh".  Personally, I prefer  ubertest. I will go ahead with the change.
 
> Is the plan to eventually remove the runfabtests.sh script? (I'd prefer a single set of scripts to maintain).

That is what I hope will happen. If all providers agree with this change. I have prepared a presentation and is planing to present it on tomorrow's ofiwg meeting.

**Comment by wckzhang** [comment]
Need to update the commit message, says that common.py provides the utility class ComplexTest. Did you also mean to add a utility class for UberTest?

**Comment by wzamazon** [comment]
> Need to update the commit message, says that common.py provides the utility class ComplexTest. Did you also mean to add a utility class for UberTest?

No, I meant replace the name "ComplexTest" with "UberTest"

**Comment by wzamazon** [comment]
We discussed this PR during the ofiwg meeting, one comment @shefty had is that there are too much code duplication, which I will address in future revision.

File: fabtests/scripts/runfabtests.py [hunk 1]
```diff
@@ -201,6 +201,11 @@ def fabtests_args_to_pytest_args(fabtests_args):
         pytest_args.append("-k")
         pytest_args.append(fabtests_args.expression)
 
+    if fabtests_args.html:
+        pytest_args.append("--html")
```
Comment (wckzhang) [inline fabtests/scripts/runfabtests.py:205]: Are these new args? Has the behavior not been implemented yet?

Comment (wzamazon) [reply fabtests/scripts/runfabtests.py:205]: Not sure what is the question?

Comment (wckzhang) [reply fabtests/scripts/runfabtests.py:205]: Ah okay, didn't realize it was a native pytest arg since it was added to fabtests_args

**Comment by shefty** [comment]
@j-xiong - do you know enough python to review this proposal?

**Comment by j-xiong** [comment]
@shefty I will take a look. I once worked with pvverbs so I think I am not totally novice :)

**Comment by wzamazon** [comment]
I just update a new revision, with the following changes:

1.  replace the term "complex" with "ubertest" to be more specific.
2. Move the handling of `-I` (iteration), `-v` (data check), `-k` (prefix mode) and `-U` (delivery complete) from each test function to class `ClientServerTest`. This reduced duplicated codes.
3. Define some commonly used parameter like `prefix_type`, `datacheck_type`, `completion_type` as parametrized fixture. This further reduced duplicated code.

@wckzhang @j-xiong @shefty Please take a look and kindly share your opinion. Thank you!

File: fabtests/pytest/pytest.ini [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[pytest]
+markers =
+    unit : unit tests runs on single instance
+    functional : runs on on two instances, not performance related
```
Comment (j-xiong) [inline fabtests/pytest/pytest.ini:4]: duplicated 'on'.

File: fabtests/pytest/pytest.ini [hunk 3]
```diff
@@ -0,0 +1,10 @@
+[pytest]
+markers =
+    unit : unit tests runs on single instance
```
Comment (j-xiong) [inline fabtests/pytest/pytest.ini:3]: runs --> run? or maybe 'single instance unit tests'?

File: fabtests/pytest/common.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        if returncode == errno.ENODATA:
+            pytest.skip("ENODATA")
+            return
+
+        if returncode == errno.ENOSYS:
+            pytest.skip("ENOSYS")
+            return
+
+    error_msg = "returncode {}".format(returncode)
+    if returncode == 124:
```
Comment (j-xiong) [inline fabtests/pytest/common.py:20]: would be helpful to add a comment about the exit value of the `timeout` command.

File: fabtests/pytest/common.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        print("stdout: ")
+        print(open(outfile).read())
+        os.unlink(outfile)
+
+        assert not timeout, "timed out"
+        check_returncode(process.returncode, self._cmdline_args.strict_fabtests_mode)
+
+class ClientServerTest:
+
+    def __init__(self, cmdline_args, executable, iteration_type=None, completion_type="transmit_complete", prefix_type="wout_prefix", datacheck_type="wout_datacheck", message_size=None):
```
Comment (j-xiong) [inline fabtests/pytest/common.py:64]: Long lines like this should be broken down. The full line is not even shown up in the review window.

File: fabtests/scripts/runfabtests.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+        if test == "quick" or test == "ubertest_quick":
+            return "quick"
+
+        if test == "all" or test == "ubertest_all":
+            return "all"
+
+        if test == "verify" or test == "ubertest_verify":
+            return "verify"
+
+        if test == "ubertest":
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply fabtests/scripts/runfabtests.py]: No reason. Fixed in new revision.

File: fabtests/pytest/pytest.ini (hunk 2 above)
Comment (wzamazon) [reply fabtests/pytest/pytest.ini]: Fixed in new revision.

File: fabtests/pytest/pytest.ini (hunk 3 above)
Comment (wzamazon) [reply fabtests/pytest/pytest.ini]: Thank you! Fixed in new revision.

File: fabtests/pytest/common.py (hunk 4 above)
Comment (wzamazon) [reply fabtests/pytest/common.py:22]: Added a comment in new revision.

File: fabtests/pytest/common.py (hunk 5 above)
Comment (wzamazon) [reply fabtests/pytest/common.py]: In new revision, I broke down this line and many other lines. I checked there are not lines longer than 120.

**Review by j-xiong — APPROVED** [review]

**Comment by wzamazon** [comment]
@j-xiong

Thank you for approving!

Is the "continuous-integration/jenkins/pr-head" failure related to this PR?

**Comment by aingerson** [comment]
@wzamazon Not related. Our cluster went down so it just can't connect. Working on it.

**Comment by j-xiong** [comment]
@wzamazon The Jenkins failure is unrelated, see the full log:

```
Branch indexing
Connecting to https://api.github.com using gh-personal access token140ce5f7b7ee4115d357ee162365cfb080b8395f (as on 12/19/2020)
Loading trusted files from base branch main at 918ea62fe35f0a5ab2b4ae956e810b0bbd18075f rather than a3170360042045496a8068a63f20e66617d07e2a
hudson.plugins.git.GitException: Command "git init /jenkinsbuild/caches/git-c0cbaff58535cb4414cff4243d9cf1f6" returned status code 128:
stdout: 
stderr: fatal: open /dev/null or dup failed: Permission denied

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2681)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2611)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2607)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommand(CliGitAPIImpl.java:1974)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$5.execute(CliGitAPIImpl.java:1046)
Caused: hudson.plugins.git.GitException: Could not init /jenkinsbuild/caches/git-c0cbaff58535cb4414cff4243d9cf1f6
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$5.execute(CliGitAPIImpl.java:1048)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.init(CliGitAPIImpl.java:352)
	at hudson.plugins.git.GitAPI.init(GitAPI.java:218)
	at jenkins.plugins.git.GitSCMFileSystem$BuilderImpl.build(GitSCMFileSystem.java:396)
	at jenkins.scm.api.SCMFileSystem.of(SCMFileSystem.java:293)
	at org.jenkinsci.plugins.workflow.multibranch.SCMBinder.create(SCMBinder.java:103)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:310)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:429)

GitHub has been notified of this commit’s build result

Finished: FAILURE
```

**Pull request merged**